File: java/src/kanzi/test/TestIntBTree.java
/*
 Copyright 2011-2013 Frederic Langlet
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 you may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limit
*/
package kanzi.test;

import java.util.Random;
import kanzi.util.IntBTree;
import kanzi.util.IntBTree.IntBTNode;


public class TestIntBTree 
{
   public static void main(String[] args)
   {
      System.out.println("Correctness Test");
      
      {
         Random random = new Random();
         IntBTree tree = new IntBTree();

         IntBTree.Callback printNode = new IntBTree.Callback()
         {
            @Override
            public void call(IntBTNode node) 
            {
               System.out.print(node.value()+", ");
            }        
         };

         for (int ii=0; ii<5; ii++)
         {
            System.out.println("\nIteration "+ii);
            int max = 0;
            int min = Integer.MAX_VALUE;

            for (int i=0; i<30; i++)
            {
               int val = 64+random.nextInt(5*i+20);
               min = (min < val) ? min : val;
               max = (max > val) ? max : val;
               tree.add(val);
               System.out.println("Add: "+val);
               System.out.print("Min/max: "+tree.min()+" "+tree.max());
               System.out.println(" Size: "+tree.size());

               if (tree.min() != min)
               {
                  System.err.println("Error: Found min="+tree.min()+", expected min="+min);
                  System.exit(1);
               }

               if (tree.max() != max)
               {
                  System.err.println("Error: Found max="+tree.max()+", expected max="+max);
                  System.exit(1);
               }
            }

            System.out.print("All nodes in reverse order: ");
            tree.scan(printNode, true);            System.out.println("");


            while (tree.size() > 0)
            {
               min = tree.min();
               max = tree.max();
               tree.remove(min);
               tree.remove(max);
               System.out.print("Remove: " + min + " " + max);
               System.out.println(" Size: " +tree.size());

               if (tree.size() > 0) 
               {
                  min = tree.min();
                  max = tree.max();
                  System.out.println("Min/max: " + min + " " + max);
               }
            }

            System.out.println("Success\n");
         }
      }
      
      System.out.println("Speed Test");
      
      {
         int iter = 10000;
         int size = 10000;
         long before1, after1;
         long before2, after2;
         long before3, after3;
         long delta1 = 0;
         long delta2 = 0;
         long delta3 = 0;
         int[] array = new int[size];
         
         Random random = new Random();
         
         for (int ii=0; ii<iter; ii++)
         {
            IntBTree tree = new IntBTree();
            
            for (int i=0; i<size; i++)
              array[i] = random.nextInt(size/2);
             
            before1 = System.nanoTime();

            for (int i=0; i<size; i++)
               tree.add(array[i]);
            
            after1 = System.nanoTime();
            delta1 += (after1 - before1);
            
            // Sanity check
            if (tree.size() != size)
            {
               System.err.println("Error: found size="+tree.size()+", expected size="+size);
               System.exit(1);
            }
            
            before2 = System.nanoTime();

            for (int i=0; i<size; i++)
               tree.remove(array[i]);
            
            after2 = System.nanoTime();
            delta2 += (after2 - before2);

            // Sanity check
            if (tree.size() != 0)
            {
               System.err.println("Error: found size="+tree.size()+", expected size="+0);
               System.exit(1);
            }

            // Recreate a 'size' array tree
            for (int i=0; i<size; i++)
               tree.add(array[i]);
            
            for (int i=0; i<size; i++)
            {
               int val = random.nextInt(size/2);
               before3 = System.nanoTime();
               tree.add(val);
               tree.remove(val);
               after3 = System.nanoTime();
               delta3 += (after3 - before3);
            }

            // Sanity check
            if (tree.size() != size)
            {
               System.err.println("Error: found size="+tree.size()+", expected size="+size);
               System.exit(1);
            }
         }

         System.out.println(size*iter+" iterations");
         System.out.println("Additions [ms]: "+(delta1/1000000L));
         System.out.println("Deletions [ms]: "+(delta2/1000000L));
         System.out.println("Additions/Deletions at size="+size+" [ms]: "+(delta3/1000000L));
      }      
   }   
}

File: go/src/kanzi/test/TestIntBTree.go
/*
Copyright 2011-2013 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"fmt"
	"kanzi/util"
	"math/rand"
	"os"
	"time"
)


func main() {
	fmt.Printf("Correctness Test\n")

	{
		tree, _ := util.NewIntBTree()
		r := rand.New(rand.NewSource(time.Now().UnixNano()))

		for ii := 0; ii < 5; ii++ {
			fmt.Printf("\nIteration %v\n", ii)
			max := 0
			min := 1 << 31

			for i := 0; i < 30; i++ {
				val := 64 + r.Intn(5*i+20)

				if val < min {
					min = val
				}

				if val > max {
					max = val
				}

				tree.Add(val)
				tMin, err1 := tree.Min()
				tMax, err2 := tree.Max()

				if err1 != nil {
					fmt.Printf("Error: %v", err1)
					os.Exit(1)
				}

				if err2 != nil {
					fmt.Printf("Error: %v", err2)
					os.Exit(1)
				}

				fmt.Printf("Add: %v\n", val)
				fmt.Printf("Min/max: %v %v\n", tMin, tMax)
				fmt.Printf("Size: %v\n", tree.Size())

				if tMin != min {
					fmt.Printf("Error: Found min=%v, expected min=%v\n", tMin, min)
					os.Exit(1)
				}

				if tMax != max {
					fmt.Printf("Error: Found max=%v, expected max=%v\n", tMax, max)
					os.Exit(1)
				}
			}

			//	fmt.Printf("All nodes in reverse order:\n")
			//	tree.Scan(printNode, true)
			println()
			fmt.Printf("All nodes in natural order\n")
			array := tree.ToArray(make([]int, tree.Size()))

			for i := range array {
				fmt.Printf("%v, ", array[i])
			}

			println()
			fmt.Printf("Size: %v\n", tree.Size())

			for tree.Size() > 0 {
				tMin, err1 := tree.Min()
				tMax, err2 := tree.Max()

				if err1 != nil {
					fmt.Printf("Error: %v\n", err1)
					os.Exit(1)
				}

				if err2 != nil {
					fmt.Printf("Error: %v\n", err2)
					os.Exit(1)
				}

				tree.Remove(tMin)
				tree.Remove(tMax)
				fmt.Printf("Remove: %v %v\n", tMin, tMax)
				array = tree.ToArray(make([]int, tree.Size()))

				for i := range array {
					fmt.Printf("%v, ", array[i])
				}

				println()
				fmt.Printf("Size: %v\n", tree.Size())

				if tree.Size() > 0 {
					tMin, err1 = tree.Min()
					tMax, err2 = tree.Max()

					if err1 != nil {
						fmt.Printf("Error: %v\n", err1)
						os.Exit(1)
					}

					if err2 != nil {
						fmt.Printf("Error: %v\n", err2)
						os.Exit(1)
					}

					fmt.Printf("Min/max: %v %v\n", tMin, tMax)
				}
			}

			fmt.Printf("Success\n\n")
		}
	}

	fmt.Printf("Speed Test\n")

	{
		iter := 5000
		size := 10000
		var before, after time.Time
		delta01 := int64(0)
		delta02 := int64(0)
		delta03 := int64(0)
		delta04 := int64(0)
		array := make([]int, size)

		for ii := 0; ii < iter; ii++ {
			tree1, _ := util.NewIntBTree()
			array[0] = 100000
			r := rand.New(rand.NewSource(time.Now().UnixNano()))

			for i := 0; i < size; i++ {
				array[i] = r.Intn(size / 2)
			}

			before = time.Now()

			for i := 0; i < size; i++ {
				tree1.Add(array[i])
			}

			after = time.Now()
			delta01 += after.Sub(before).Nanoseconds()
			before = time.Now()

			for i := 0; i < size; i++ {
				tree1.Contains(array[size-1-i])
			}

			after = time.Now()
			delta04 += after.Sub(before).Nanoseconds()

			// Sanity check
			if tree1.Size() != size {
				fmt.Printf("Error: Found size=%v, expected size=%v\n", tree1.Size(), size)
				os.Exit(1)
			}

			before = time.Now()

			for i := 0; i < size; i++ {
				tree1.Remove(array[i])
			}

			after := time.Now()
			delta02 += after.Sub(before).Nanoseconds()

			// Sanity check
			if tree1.Size() != 0 {
				fmt.Printf("Error: Found size=%v, expected size=%v\n", tree1.Size(), 0)
				os.Exit(1)
			}

			// Recreate a 'size' array tree
			for i := 0; i < size; i++ {
				tree1.Add(array[i])
			}

			for i := 0; i < size; i++ {
				val := rand.Intn(size / 2)
				before = time.Now()
				tree1.Add(val)
				tree1.Remove(val)
				after = time.Now()
				delta03 += after.Sub(before).Nanoseconds()
			}

			// Sanity check
			if tree1.Size() != size {
				fmt.Printf("Error: Found size=%v, expected size=%v\n", tree1.Size(), size)
				os.Exit(1)
			}
		}

		fmt.Printf("%v iterations\n", size*iter)
		fmt.Printf("Additions [ms]: %d\n", delta01/1000000)
		fmt.Printf("Deletions [ms]: %d\n", delta02/1000000)
		fmt.Printf("Contains  [ms]: %d\n", delta04/1000000)
		fmt.Printf("Additions/Deletions at size=%v [ms]: %d\n", size, delta03/1000000)
	}
}
